### ISSUE TITLE
wip commit errors when adding file deleted in the working tree 

### ISSUE THREAD

**Issue by kyleam** [magit/magit#4040]
When we call `magit-wip-commit-worktree` after staging a deleted file, the `(magit-call-git "add" "--" files)` call inside fails because the file no longer exists in the working tree.  I don't think the `git add` call failing is problematic by itself, but there are two issues:

  * A user seeing the message (in the minibuffer or status buffer) about the failure would reasonably be concerned that there is an issue.
  * If there are other, non-deleted files that come after the deleted file in the pathspec, they will not be added to the index.

The best I can think of for dealing with this is something like

```diff
diff --git a/lisp/magit-wip.el b/lisp/magit-wip.el
index 6fcdf19b..e3d36897 100644
--- a/lisp/magit-wip.el
+++ b/lisp/magit-wip.el
@@ -283,7 +283,10 @@ (defun magit-wip-commit-worktree (ref files msg)
          (parent (magit-wip-get-parent ref wipref))
          (tree (magit-with-temp-index parent "--reset"
                  (if files
-                     (magit-call-git "add" "--" files)
+                     (let ((magit-process-raise-error t)
+                           (magit-process-display-mode-line-error nil))
+                       (ignore-errors
+                         (magit-call-git "add" "--ignore-errors" "--" files)))
                    (magit-with-toplevel
                      (magit-call-git "add" "-u" ".")))
                  (magit-git-string "write-tree"))))
```

That should make the error much less prominent, though the user may of course still see the failing command in the process buffer and be confused.  The `--ignore-errors` takes care of the second issue.

Any other ideas?

**Comment by tarsius** [comment]
> Any other ideas?

It's a bit hackish, but that seems okay given that we do it only once. You should probably add a comment.

If we start doing this more often, then we should come up with a more explicit way. Maybe a macro that also comes with a doc-string, or even a new `magit-*git*` function. But implementing something like that now seems a bit premature.

**Comment by kyleam** [comment]
Thanks for your input.

> It's a bit hackish,

It is indeed.

> but that seems okay given that we do it only once. You should probably add a comment.

Yep, the suggested change should/would be accompanied by a comment.

**Comment by kyleam** [comment]
> the `(magit-call-git "add" "--" files)` call inside fails because the file no longer exists in the working tree

In case anyone was scratching their head over my incorrect/incomplete description: no longer exists in the working tree _and_ does not exist in the temporary index we're trying to add it to.

---

I didn't get a chance to revisit this tonight, but it did occur that me that the `ignore-errors` should be restricted to `magit-git-error`.

**Comment by kyleam** [comment]
It turns out my initial assessment was wrong in two ways:

  * I talk about the problem of other files not being added if they come _after_ the problematic deleted file in the pathspec, but the order doesn't actually matter.  No files are added on failure.
  * `add --ignore-errors` doesn't actually let us add more files in the case of this particular failure so using it doesn't solve the issue above.  (I had tried it out before concluding it did, but I must've confused myself.)

Anyway, I've pushed an alternative approach in 7b5a0333 (km/wip-wtree-deleted-error).  It avoids the error altogether by using `update-index`.  I'll run with that locally tomorrow, and then push it to master tomorrow night if there are no objections.
